### PR TITLE
Draft optics implementation

### LINKED ISSUE #3: Optics

### PR DESCRIPTION
**Pull request by AriFordsham**
Initial draft implementation.

### CODE REVIEW

**Comment by AriFordsham** [comment]
What obstacles are there to 'efficiently?'

I'm planning on implementing profunctor optics in a way that makes sense for Plutarch lenses.

- As much work should be done in Haskell, at Plutarch compile time.
- I need to ensure not to introduce superfluous `pcon`/`pmatch` into the body of combinators and lenses themselves.
- Is there a risk of redundant `pcon`/`pmatch` in a composition of optics, which the composition function needs to deal with? It doesn't seem obvious to me that there is.
- The general problem of redundant matches might be unsolvable in library, and may need an optimizer pass.

Are there any efficiency considerations I am missing?

**Comment by L-as** [comment]
Consider combining `Lens' a b` and `Lens' a c` into `Lens' a (b, c)`. How do you this efficiently?
Relying on the backend to optimise it is not good design.

**Comment by AriFordsham** [comment]
- Optics for `EPair` and `EEither`
- type aliases `PLens` and `PPrism`
- typeclass `PChoice` for `PPrism`
- some useful combintors in `Core`

**Comment by AriFordsham** [comment]
We have optics!

- Plutarch-free profunctor and optics stacks based on the `Cont` monad - essentially replacing every `a -> b` with `a -> Cont r b`
- Optics for `EPair` and `EEither`. These essentially use Haskell `Either` in Plutarch!

**Comment by L-as** [comment]
Why change the warnings?
Make sure to use GHC 9.2.4. You can do `nix develop` to get the correct GHC.

**Comment by AriFordsham** [comment]
Happy to change that back. I use `partial-type-signatures` warnings with HLS quite heavily in my workflow.
`shell.nix` on `master` has `ghc923`, which is what I'm using.

File: Plutarch/CPS/Profunctor.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+module Plutarch.CPS.Profunctor where
+
+import Control.Monad.Cont
+import Data.Tuple
+
+newtype CStar r f a b = CStar { runCStar :: a -> Cont r (f b) }
+  
+class CProfunctor r p where
+  cdimap :: (a -> Cont r b) -> (c -> Cont r d) -> p b (Cont r c) -> p a (Cont r d)
```
Comment (L-as) [inline Plutarch/CPS/Profunctor.hs:11]: Maybe this should still be `p b c` and `p a d`.

Comment (L-as) [reply Plutarch/CPS/Profunctor.hs:11]: Then `CProfunctor r ->` would not be implementable, but you could make a newtype where it is implementable.

**Comment by L-as** [comment]
e.g.?

**Comment by AriFordsham** [comment]
Not having `r` on the right of an arrow breaks a lot of inference and requires many type applications. In addition, something hard broke, I can't remember what. It seemed fixable, but not worth the trouble. I'm happy to try again if you think it's worth it.

**Comment by L-as** [comment]
In general, type applications are a code smell, along with AllowAmbiguousTypes IMO.

**Comment by AriFordsham** [comment]
`PConstrained` implemented

**Comment by AriFordsham** [comment]
- Completed `ptraverse` for types with one parameter
- Implemented `plist` in terms of `ptraverse`

We won't really know if this works until we have a backend to test with.

**Comment by L-as** [comment]
Amazing! Perhaps we should work on getting a backend to work again.

**Comment by L-as** [comment]
Plutarch has changed too much for this to be mergible.

**Pull request closed without merging**